quadra/utils/vit_explainability.py

```diff
@@ -0,0 +1,283 @@
+            )
+        else:
+            self.model = model  # type: ignore[assignment]
+
+        self.discard_ratio = discard_ratio
+        self.f_hook_handles: List[torch.utils.hooks.RemovableHandle] = []
+        self.b_hook_handles: List[torch.utils.hooks.RemovableHandle] = []
+        for name, module in self.model.named_modules():
+            for layer_name in attention_layer_names:
+                if layer_name in name:
+                    self.f_hook_handles.append(module.register_forward_hook(self.get_attention))
+                    self.b_hook_handles.append(module.register_backward_hook(self.get_attention_gradient))
+        self.attentions: List[torch.Tensor] = []
+        self.attention_gradients: List[torch.Tensor] = []
+        # Activate gradients
+        blocks_list = [x.split("blocks")[1].split(".attn")[0] for x in attention_layer_names]
+        for name, module in model.named_modules():
+            for p in module.parameters():
+                if "blocks" in name and any(x in name for x in blocks_list):
+                    p.requires_grad = True
+
+    def get_attention(
+        self, module: torch.nn.Module, inpt: torch.Tensor, out: torch.Tensor  # pylint: disable=W0613
+    ) -> None:
+        """Hook to return attention.
+
+        Args:
+            module: Torch module
+            inpt: Input tensor
+            out: Output tensor, in this case the attention
+        """
+        self.attentions.append(out.detach().clone().cpu())
+
+    def get_attention_gradient(
+        self, module: torch.nn.Module, grad_input: torch.Tensor, grad_output: torch.Tensor  # pylint: disable=W0613
+    ) -> None:
+        """Hook to return attention.
+
+        Args:
+            module: Torch module
+            grad_input: Gradients' input tensor
+            grad_output: Gradients' output tensor, in this case the attention
+        """
+        self.attention_gradients.append(grad_input[0].detach().clone().cpu())
+
+    def __call__(self, input_tensor: torch.Tensor, targets_list: List[int]) -> np.ndarray:
+        """Called when the class instance is used as a function.
+
+        Args:
+            input_tensor: Model's input tensor
+            targets_list: List of targets. If None, argmax is used
+
+        Returns:
+            out: Batch of output masks
+        """
+        self.attentions.clear()
+        self.attention_gradients.clear()
+
+        self.model.zero_grad(set_to_none=True)
+        self.model.to(input_tensor.device)
+        output = self.model(input_tensor).cpu()
+
+        class_mask = torch.zeros(output.shape)
+        if targets_list is None:
+            targets_list = output.argmax(dim=1)
+        class_mask[torch.arange(output.shape[0]), targets_list] = 1
+        loss = (output * class_mask).sum()
+        loss.backward()
+        out = grad_rollout(self.attentions, self.attention_gradients, self.discard_ratio)
+
+        return out
+
+
+class ClassifierPytorchWrapper(torch.nn.Module):
```

Comment:
Is this working from anything different than a logistic regression? 

If not I would check it before

Reply:
Thinking about it, maybe it is suitable for sklearn.linear_model class of models only. I will verify today